File: js-pr/56/javascript/code-quality-readability-maintainability/url_builder_46.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
function constructUrl(baseUrl, path = null, queryParams = null) {
    const urlParts = [baseUrl.replace(/\/$/, '')];
    
    if (path) {
        const cleanPath = path.replace(/^\//, '');
        urlParts.push(cleanPath);
    }
    
    let constructedUrl = urlParts.join('/');
    
    if (queryParams && Object.keys(queryParams).length > 0) {
        const queryString = buildQueryString(queryParams);
        constructedUrl = `${constructedUrl}?${queryString}`;
    }
    
    return constructedUrl;
}

function buildQueryString(params) {
    return Object.entries(params)
        .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
        .join('&');
}
// {/fact}

File: python-pr/52/python/code-quality-readability-maintainability/url_builder_46.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
def construct_url(base_url, path=None, query_params=None):
    url_parts = [base_url.rstrip('/')]
    
    if path:
        clean_path = path.strip('/')
        url_parts.append(clean_path)
    
    constructed_url = '/'.join(url_parts)
    
    if query_params:
        query_string = _build_query_string(query_params)
        constructed_url = f"{constructed_url}?{query_string}"
    
    return constructed_url

def _build_query_string(params):
    param_pairs = [f"{key}={value}" for key, value in params.items()]
    return '&'.join(param_pairs)
# {/fact}